libs/libskynet/src/skylinkBitfield.ts

```diff
@@ -0,0 +1,166 @@
+// parseSkylinkBitfield parses a skylink bitfield and returns the corresponding
+// version, offset, and fetchSize.
+function parseSkylinkBitfield(skylink: Uint8Array): [bigint, bigint, bigint, string | null] {
+  // Validate the input.
+  if (skylink.length !== 34) {
+    return [0n, 0n, 0n, "provided skylink has incorrect length"];
+  }
+
+  // Extract the bitfield.
+  let bitfield = new DataView(skylink.buffer).getUint16(0, true);
+
+  // Extract the version.
+  const version = (bitfield & 3) + 1;
+  // Only versions 1 and 2 are recognized.
+  if (version !== 1 && version !== 2) {
+    return [0n, 0n, 0n, "provided skylink has unrecognized version"];
+  }
+
+  // If the skylink is set to version 2, we only recognize the link if
+  // the rest of the bits in the bitfield are empty.
+  if (version === 2) {
+    if ((bitfield & 3) !== bitfield) {
+      return [0n, 0n, 0n, "provided skylink has unrecognized version"];
+    }
+    return [BigInt(version), 0n, 0n, null];
+  }
+
+  // Verify that the mode is valid, then fetch the mode.
+  bitfield = bitfield >> 2;
+  if ((bitfield & 255) === 255) {
+    return [0n, 0n, 0n, "provided skylink has an unrecognized version"];
+  }
+  let mode = 0;
+  for (let i = 0; i < 8; i++) {
+    if ((bitfield & 1) === 0) {
+      bitfield = bitfield >> 1;
+      break;
+    }
+    bitfield = bitfield >> 1;
+    mode++;
+  }
+  // If the mode is greater than 7, this is not a valid v1 skylink.
+  if (mode > 7) {
+    return [0n, 0n, 0n, "provided skylink has an invalid v1 bitfield"];
+  }
+
+  // Determine the offset and fetchSize increment.
+  const offsetIncrement = 4096 << mode;
+  let fetchSizeIncrement = 4096;
+  let fetchSizeStart = 0;
+  if (mode > 0) {
+    fetchSizeIncrement = fetchSizeIncrement << (mode - 1);
+    fetchSizeStart = (1 << 15) << (mode - 1);
```

Comment:
I'm totally lost here at this point without explanation, if this function is copied from skyd go implementation maybe a link to documented code would help - the question is
1. is the code reader suppose to understand the code and why it works like that from reading it
2. should we assume that the code is like this because it's a translated go code that we can read elsewhere and we're just not documenting stuff in 2 places

Reply:
it's ported from go code and required to match the protocol spec, I added a link at the top of the file.